compiler/inference/helpers.py

```diff
@@ -0,0 +1,141 @@
+    - For all accelerators whose state val changed in *at least
+      one* of the branches
+    - And for all accelerator states that got introduced in *both*
+      branches
+    """
+    new_state: dict[str, tuple[SSAValue, SSAValue]] = {}
+
+    # for every key in the old state, find out if it changed
+    for k in old_state:
+        # get new vals (or None if dropped)
+        new_vals = (
+            if_state.pop(k, None),
+            else_state.pop(k, None),
+        )
+        # drop val if it is invalidated on one side
+        if any(v is None for v in new_vals):
+            continue
+        # if no val changed
+        if all(v == old_state[k] for v in new_vals):
+            continue
+        # pyright can't infer that we actually checked the argument type:
+        new_state[k] = new_vals  # pyright: ignore[reportArgumentType]
+
+    # check for states that are present in both branches
+    for k in if_state:
+        if k not in else_state:
+            continue
+        # add them to the new dict
+        new_state[k] = (if_state[k], else_state[k])
+
+    return new_state
+
+
+def find_all_acc_names_in_region(reg: Region) -> set[str]:
+    """
+    Walk a region and return a set of accelerator names are set up in that region.
+    """
+    acs: set[str] = set()
+    for op in reg.walk():
+        if isinstance(op, accfg.SetupOp):
+            acs.add(op.accelerator.data)
+    return acs
+
+
+def find_existing_block_arg(block: Block, accel: str) -> BlockArgument | None:
+    """
+    Inspect a block for block arguments of the correct accfg.StateType type, return the block arg if found.
+    """
+    for arg in block.args:
+        if isinstance(arg.type, accfg.StateType) and arg.type.accelerator.data == accel:
+            return arg
+    return None
+
+
+def iter_ops_range(
+    start_op: Operation | None, end_op: Operation | None
+) -> Generator[Operation, None, None]:
+    """
+    Create an iterator that goes from start_op (inclusive) to end_op (exclusive).
+
+    If start_op is None, iterate from the start of the block.
+
+    If end_op is None, iterate to the end of the block (inclusive).
+
+    Both ops cannot be none.
+
+    If provided, both start and end op must be in the same block.
+    """
+
+    if start_op is None:
+        assert end_op is not None, "Can't have both start and end_op be None!"
+        parent = end_op.parent_block()
+        assert parent is not None, "Provided operations must have parent blocks!"
+        start_op = parent.first_op
+
+    assert start_op is not None  # just for pyright
+
+    if end_op is not None:
+        assert (
+            start_op.parent_block() is end_op.parent_block()
+        ), "Cannot iterate between operations not within the same block!"
+
+    yield start_op
+    while start_op.next_op is not end_op:
+        assert start_op.next_op is not None  # just for pyright
+        yield start_op.next_op
+        start_op = start_op.next_op
```

Comment:
This is a neat helper that I might upstream to xDSL at some point. It's new in this PR